[DC-1024] Remove name field from criteria

evansuslovich opened the pull request:
### Jira Ticket: https://broadworkbench.atlassian.net/browse/DC-1024

<!-- ### Dependencies --->
<!-- Include any dependent tickets and describe the relationship. Include any relevant Jira tickets. --->

## Summary of changes:
<!--Please give an abridged version of the ticket description here and/or fill out the following fields.-->

`DataRepo` adjustments: 
- Removed `name` field from `Criteria`
- renamed `getCounts` function to `getSnapshotBuilderCount`
- Removed `id` from `SnapshotBuilderFeatureValueGroup`
- Renamed `AnyCriteria` to `AnySnapshotBuilderCriteria`
- Created`SnapshotAccessRequestResponse` type
- Removed `convertDatasetParticipantCountRequest` from `convertDatasetAccessRequest`
   - These conversion will be made before the request is passed in.

Types moved from `DatasetBuilderUtils` to`DataRepo` (and): 
- `DatasetBuilderType`
- `SnapshotBuilderParentConcept`
- Renamed `GetConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
- Renamed `GetConceptHierarchyResponse `to `SnapshotBuilderGetConceptHierarchyResponse `
- Renamed `SearchConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
- Renamed `DatasetParticipantCountResponse` to `SnapshotBuilderCountResponse`
- Renamed `DatasetAccessRequestApi` to `DatasetAccessRequestApi`
- Renamed `DatasetRequestApi` to `SnapshotBuilderRequest`
- Renamed `ValueSetApi` to `SnapshotBuilderFeatureValueGroup`
- Renamed `CohortApi` to `SnapshotBuilderCohort`
- Renamed `CriteriaGroupApi` to `SnapshotBuilderCriteriaGroup`
- Renamed `ProgramDataListCriteriaApi`  to `SnapshotBuilderProgramDataListCriteria`
- Renamed `ProgramDataRangeCriteriaApi` to `SnapshotBuilderProgramDataRangeCriteria`
- Renamed `DomainCriteriaApi` to `SnapshotBuilderDomainCriteria`
- Renamed `CriteriaApi` to `SnapshotBuilderCriteria`, removed `name` and `count`

Moved from `DataRepo` to `DatasetBuilderUtils`: 
- `PrepackagedConceptSet`

Types created in `DatasetBuilderUtils`:
- `BuilderSettings`
- `DatasetModel`
- Renamed `DomainCriteria `to `ProgramDomainCriteria`


### What
- Name isn't required in Swagger
- Cleaning up types to follow `jade-data-repo` naming

### Why
- Only pass values that are required in the backend
- Tech debt

### Testing strategy
<!-- Note that changes impacting components in Storybook stories can be viewed at
https://www.chromatic.com/library?appId=65fc89c9335768720ff8605a&branch=<branch>
The branch corresponding to this PR is selected, and changes can be reviewed by commit. --->

- [ ] <!-- Test case 1 -->
- [ ] Unit Tests

<!-- ### Visual Aids -->
<!-- https://support.apple.com/guide/quicktime-player/record-your-screen-qtp97b08e666/mac -->

evansuslovich committed:
remove name

evansuslovich committed:
moving data types to DataRepo

evansuslovich committed:
moved over types to DataRepo

evansuslovich committed:
cleaning issues

evansuslovich committed:
trying a different endpoint

pshapiro4broad commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -81,6 +78,60 @@ export type SnapshotBuilderSettings = {
+  // This is the ID for either the domain or the program data option
+  id: number;
+  kind: SnapshotBuilderOptionTypeNames;
+  count?: number;
```
Can this field be removed? If this object represents the criteria that the UI is passing to the API, then I don't think it will ever have a value. Looking at the definition of `SnapshotBuilderCriteria` in TDR, there is no `count` field.

evansuslovich replied:
Yes, I agree, I'll include this issue in the `DataRepo` / `DatasetBuilderUtils` clean up ticket. I opened a ticket some time ago to look at the types in swagger and identify what should be required / removed. 
https://broadworkbench.atlassian.net/browse/DC-961

pshapiro4broad commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -81,6 +78,60 @@ export type SnapshotBuilderSettings = {
+  // This is the ID for either the domain or the program data option
+  id: number;
+  kind: SnapshotBuilderOptionTypeNames;
+  count?: number;
```
I also wonder about `SnapshotBuilderDomainOption` fields `conceptCount` and `participantCount` too. It's not part of this PR, but at some point we should revisit them as I think they should always be present in the settings data.

pshapiro4broad commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -81,6 +78,60 @@ export type SnapshotBuilderSettings = {
   datasetConceptSets?: PrepackagedConceptSet[];
 };
 
+export interface Criteria {
```
Right now we have a mix of API names that match the Java types and some that don't. For example we have `SnapshotBuilderConcept`, which does match, and `Criteria`, which doesn't (it would be `SnapshotBuilderCriteria`, I think). Since these names now clash with the UI types and need to be aliased when imported, maybe they should be changed to use the Java type names. If you agree that this is a good idea, can you make a JIRA ticket for it?

s-rubenstein replied:
I would very much agree here - in general I think the types in `DataRepo.ts` should match the Java type names where possible.

pshapiro4broad commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -2,17 +2,14 @@ import * as _ from 'lodash/fp';
   convertDatasetParticipantCountRequest,
-  DatasetAccessRequest,
-  DatasetAccessRequestApi,
+  DatasetAccessRequest as DatasetAccessRequestUtils,
```
This seems like the wrong name to me. It's true that this type is defined in a `Utils` file, but its meaning isn't `Utils` related. If it's the API type, then it should be defined in this file instead. Also there's no other `DatasetAccessRequest` in scope so no aliasing is needed in either case.

evansuslovich replied:
I agree that type isn't `Utils` related, I wasn't ready with my PR so I should have left this as a draft. However, there is a `DatasetAccessRequest` in scope here https://github.com/DataBiosphere/terra-ui/blob/25f7a33363fb902fbbc4e3553e14cbbc1574e3bf/src/libs/ajax/DataRepo.ts#L129

pshapiro4broad reviewed (approved):
Looks OK, I think it's good to reorganize the types like this, but there's one case where I think the migration isn't done yet.

evansuslovich committed:
import

evansuslovich committed:
moving DatasetParticipantCountResponse and DatasetParticipantCountRequest back to DatasetBuilderUtils

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
use only one DataRepo import

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
move count request and response

evansuslovich committed:
move count request and response

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

rjohanek commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -142,11 +195,28 @@ export interface JobModel {
     details: (include?: DatasetInclude[]) => Promise<DatasetModel>;
     roles: () => Promise<string[]>;
-    createSnapshotRequest(request: DatasetAccessRequest): Promise<DatasetAccessRequestApi>;
+    createSnapshotRequest(request: DatasetAccessRequestUtils): Promise<DatasetAccessRequest>;
```
Why do we want to change this from the API type? Or is this just a renaming change?

evansuslovich replied:
This is a renaming change

evansuslovich replied:
The request is `DatasetAccessRequestUI` (from `DatasetBuilderUtils`) and it's promised the type `DatasetAccessRequest` (in `DataRepo). This should be `DatasetAccessResponse` but I believe from what I see in our tests, what's passed in is the same as the return.

evansuslovich replied:
I'll add this to the refactoring ticket, this is a bit confusing. I like the convention of having `Response` at the end.

rjohanek commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -81,6 +80,60 @@ export type SnapshotBuilderSettings = {
+  count?: number;
+}
+
+export interface DomainCriteria extends Criteria {
```
are these all still api types, just moved to a new file? I find it confusing to tell the difference between the UI types and API types if they have the same name (instead of for example DomainCriteriaApi). I like that you moved them over to this file, but I also like the distinct Api names. What are your thoughts on this?

evansuslovich replied:
You're correct, these are all api types moved to a new file. There are differences with the contents of the UI and API types: 

For example, `CriteriaGroup` looks completely similar here: 

<img width="1068" alt="Screenshot 2024-04-26 at 2 34 46 PM" src="https://github.com/DataBiosphere/terra-ui/assets/63474660/392e585a-0cd0-4760-879b-6e77ad123f08">

But, as we beginning climbing up to`AnyCriteria`, and then `DomainCriteria` we can see that they have different specifications: 
<img width="1066" alt="Screenshot 2024-04-26 at 2 35 40 PM" src="https://github.com/DataBiosphere/terra-ui/assets/63474660/c7f813b6-a7e4-4b26-b4ac-bd48795d0687">


I agree with you that including the `Api` is helpful, that's why whenever they're used outside of `DataRepo` we use an alias. ex: 

`import { DomainCriteria as DomainCriteriaApi } from 'DataRepo'`

evansuslovich replied:
I put the API and UI types in the same order that way, it's a lot easier to compare, I found it really difficult to tell the differences when they were in one file. Does this make sense?

rjohanek replied:
Okay, thanks for your explanation! I guess my suggestion would be to not just rely on aliasing but to have the names actually be different when the types are different. Or maybe i just need to understand, what was the reason in removing the 'Api' from the api types when they differ from the UI types?

evansuslovich replied:
Removing `Api` from the type names were based off moving from `DatasetBuilderUtils`, where the UI types live, to`DataRepo` where the backend types live. Having `Api` in the type name was useful when they were in `DatasetBuilderUtils` because it was a point of differentiation. I don't want my bias to get in the way, but if you think removing `Api` adds extra confusion, let's chat about this decision with the rest of the team?

rjohanek replied:
Okay, sounds good. If you and others don't find it confusing then its fine with me!

evansuslovich replied:
Ok sweet. When I mean bias, I also mean the bias of writing this code. The a

snf2ye replied:
Did we get a chance to discuss this further?

s-rubenstein replied:
I think some of the confusion goes away if we are consistent with what types go where, and what names we use. If the ones in `DataRepo.ts` always match the API names and type definitions, and the ones in `DatasetBuilderUtils.ts` are our custom types, then we can rely on that to tell the difference. And then only alias when it doesn't add more confusion.

evansuslovich committed:
add more distinct type

evansuslovich committed:
Merge branch 'es/dc-1024-remove-name-field-from-criteria' of https://github.com/DataBiosphere/terra-ui into es/dc-1024-remove-name-field-from-criteria
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

s-rubenstein commented on src/dataset-builder/ConceptSearch.ts:
```diff
@@ -8,8 +8,13 @@ import { TextInput, withDebouncedChange } from 'src/components/input';
+import { formatCount, HighlightConceptName } from 'src/dataset-builder/DatasetBuilderUtils';
+import {
+  DataRepo,
+  GetConceptsResponse,
```
Since this object in the yaml is called `SnapshotBuilderGetConceptsResponse`, did you think about renaming it that? If so, you could do the import here as `SnapshotBuilderGetConceptsResponse as GetConceptsResponse` to maintain the conciseness. (And, if you wanted, you could also alias `DomainOption`?)

snf2ye commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -142,11 +195,28 @@ export interface JobModel {
   class_name?: string;
 }
 
+export interface GetConceptsResponse {
+  result: SnapshotBuilderConcept[];
+}
+
+export interface GetConceptHierarchyResponse {
+  readonly result: SnapshotBuilderParentConcept[];
+}
+
+export interface SearchConceptsResponse {
+  result: SnapshotBuilderConcept[];
+}
+
+export interface SnapshotBuilderParentConcept {
+  parentId: number;
+  children: SnapshotBuilderConcept[];
+}
+
 export interface DataRepoContract {
   dataset: (datasetId: string) => {
     details: (include?: DatasetInclude[]) => Promise<DatasetModel>;
     roles: () => Promise<string[]>;
-    createSnapshotRequest(request: DatasetAccessRequest): Promise<DatasetAccessRequestApi>;
+    createSnapshotRequest(request: DatasetAccessRequestUI): Promise<DatasetAccessRequest>;
```
This translation from the UI to the API type is confusing to me. Why can't we use the DatasetAccessRequest in both spots?

s-rubenstein replied:
I think one of the big reasons is that we have wanted additional information on the UI that doesn't make sense to send up to the API/have the API accept. However, it is weird that that is at the request level. I would expect the request object we build up to always match the API version, and do the conversions as we build up the request.

snf2ye replied:
Gotcha, make sense. I'm curious -  @evansuslovich  - What happens if you remove the DatasetAccessRequestUI alias here?

evansuslovich replied:
Yes, if we remove `DatasetAccessRequestUI` and replace it with `DatasetAccessRequest` there will be issues because there are extra fields in the UI type, we can see that there is a `convertDatasetAccessRequest` which converts the `cohorts and `valueSets` with the `convertCohort` and `convertValueSet` functions.

```
export const convertDatasetAccessRequest = (datasetAccessRequest: DatasetAccessRequest) => {
  return {
    name: datasetAccessRequest.name,
    researchPurposeStatement: datasetAccessRequest.researchPurposeStatement,
    datasetRequest: {
      cohorts: _.map(convertCohort, datasetAccessRequest.datasetRequest.cohorts),
      conceptSets: datasetAccessRequest.datasetRequest.conceptSets,
      valueSets: _.map(convertValueSet, datasetAccessRequest.datasetRequest.valueSets),
    },
  };
};
```

s-rubenstein replied:
I think we could use the convert methods earlier, and thus do without the extra UI type, since we don't need that anywhere. That would allow us to do without the `DatasetAccessRequestUI` type, since we never construct it with the extra fields in the first place.

evansuslovich replied:
Yes, I agree. I've removed all the types and I've converted at every instance of calling DataRepo functions

s-rubenstein commented on src/dataset-builder/DatasetBuilderUtils.test.ts:
```diff
@@ -1,35 +1,35 @@
-  ValueSetApi,
 } from 'src/dataset-builder/DatasetBuilderUtils';
 import {
+  AnyCriteria as AnyCriteriaApi,
```
In the yaml, this is called `SnapshotBuilderCriteria` (I think?), how would you feel about naming it as that? Or, if we need to have a situation where it is:
```
interface SnapshotBuilderCriteria {}
interface SnapshotBuilderDomainCriteria extends SnapshotBuilderCriteria {}
...
type {{NAME}} = SnapshotBuilderDomainCriteria | SnapshotBuilderProgramDataListCriteria | SnapshotBuilderProgramDataRangeCriteria;
```
then I would consider naming it `AnySnapshotBuilderCriteria`. How does that sound to you?

evansuslovich replied:
This sounds good!

s-rubenstein reviewed (commented):
I think this is a good improvement. I would love to see the change to name the `DataRepo.ts` names after the Java types more fully. This should make things much more clean. And then, after that, how would you feel making sure that `DataRepo.ts` doesn't import from utils? I feel like including the conversion and logic in the `DataRepo.ts` file introduces a lot of confusion, and it might be cleaner to just have that file fully represent the state of the API, with no logic. And push the logic and UI conversions into the `DatasetBuilderUtils`.

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
renaming files to mimick jade-data-repo types

evansuslovich committed:
Merge branch 'es/dc-1024-remove-name-field-from-criteria' of https://github.com/DataBiosphere/terra-ui into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
removed import

evansuslovich committed:
removed import

evansuslovich committed:
PrepackagedConceptSet can be replaced with SnapshotBuilderStatsetConceptSet because they're the same

evansuslovich committed:
remove imports

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
replace getCounts with getSnapshotBuilderCount

evansuslovich committed:
rename getCounts --> getSnapshotBuilderCount

evansuslovich commented on src/dataset-builder/CohortEditor.test.ts:
```diff
@@ -99,7 +99,7 @@ describe('CohortEditor', () => {
       dataset: (_datasetId) =>
         ({
-          getCounts: () =>
+          getSnapshotBuilderCount: () =>
```
renamed `getCounts` function to `getSnapshotBuilderCount`

evansuslovich commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -53,32 +51,69 @@ export interface SnapshotBuilderProgramDataRangeOption extends SnapshotBuilderPr
-  values: string[];
-};
-
-export interface DomainConceptSet extends ConceptSet {
```
This is used in the UI only, this doesn't have to belong in `DataRepo`

evansuslovich commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -53,32 +51,69 @@ export interface SnapshotBuilderProgramDataRangeOption extends SnapshotBuilderPr
 }
 
-export type SnapshotBuilderFeatureValueGroup = {
-  id: number;
```
This is moved higher in the file, and `id` is removed

evansuslovich commented on src/dataset-builder/DatasetBuilderUtils.ts:
```diff
@@ -21,72 +27,17 @@ export interface Criteria {
   conceptId: number;
   conceptName: string;
-  option: SnapshotBuilderDomainOption;
+  option: DomainOption;
```
To include `conceptCount` and `participantCount` in the UI type we created a new type called `DomainOption` with these values

pshapiro4broad replied:
This may have been due to my suggestion, but since these are used on the details page I think they should be required in the API types too, and if they are, we don't need another type here.

Also, looking at the code that uses these fields (`TileDisplay` in `DatasetBuilderDetails.ts`) it is still referring to the API types, not these new types. I think the reason why this isn't flagged as an error is because these fields are optional in the downstream type (`DomainDisplayProps.displayInformation`). Making these types non-optional should expose the issue.

evansuslovich commented on src/dataset-builder/DatasetBuilderUtils.ts:
```diff
@@ -209,13 +171,6 @@ export type DatasetParticipantCountRequest = {
   cohorts: Cohort[];
 };
 
-export type DatasetParticipantCountResponse = {
```
This is used in `DataRepo` only

evansuslovich committed:
renamed from DomainCriteria to ProgramDomainCriteria to follow pattern of Criterias' starting with Program

evansuslovich committed:
remove alias of DatasetAccessRequestUI to DatasetAccessRequest and DatasetParticipantCountRequest from DatasetBuilderUtils to DataRepo

evansuslovich committed:
move back DatasetParticipantCountRequest because I'd rather import this than 'Cohort' and create a local DatasetBuilderType so I don't import this from DatasetBuilderUtilsg

evansuslovich commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -142,15 +180,46 @@ export interface JobModel {
+  result: SnapshotBuilderParentConcept[];
+}
+
+export interface SnapshotAccessRequestResponse {
```
Reference `SnapshotAccessRequestResponse` from `jade-data-repo`, what are ways `UUID` has been implemented in the past?

s-rubenstein replied:
Just as string - the UI doesn't often do things with IDs, so we don't really need the more strict typings

evansuslovich committed:
use DatasetBuilderUtils type and use UUID

evansuslovich committed:
add expected return types to 'convert' functions, remove convert functions from DataRepo

evansuslovich committed:
remove return from searchConcepts

evansuslovich committed:
adding comments, moving types around for clarity, removed DatasetBuilderType from DatasetBuilderUtils because I'd rather have it in DataRepo

evansuslovich committed:
renamed AnyCriteria to AnySnapshotBuilderCriteria

evansuslovich commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -50,37 +41,90 @@ export interface SnapshotBuilderProgramDataRangeOption extends SnapshotBuilderPr
-  concept: SnapshotBuilderConcept;
 }
 
-export type PrepackagedConceptSet = ConceptSet;
```
This also doesn't need to be in `DataRepo`, as its only used in the UI

evansuslovich committed:
removing aliasing

evansuslovich committed:
include expected types, include convert

pshapiro4broad commented on src/libs/ajax/DataRepo.ts:
```diff
@@ -50,37 +41,90 @@ export interface SnapshotBuilderProgramDataRangeOption extends SnapshotBuilderPr
   programDataOptions: (SnapshotBuilderProgramDataListOption | SnapshotBuilderProgramDataRangeOption)[];
   featureValueGroups: SnapshotBuilderFeatureValueGroup[];
-  datasetConceptSets?: PrepackagedConceptSet[];
+  datasetConceptSets?: SnapshotBuilderDatasetConceptSet[];
```
I think this can be made non-optional. Although I see that it's not yet marked as `required` in the openapi spec, so I guess it can wait until that's changed.

pshapiro4broad commented on src/dataset-builder/DatasetBuilderUtils.ts:
```diff
@@ -114,54 +69,60 @@ export interface Cohort extends DatasetBuilderType {
+export interface DomainOption extends SnapshotBuilderOption {
+  kind: 'domain';
+  root: SnapshotBuilderConcept;
+  conceptCount?: number;
+  participantCount?: number;
```
If these aren't non-optional in the API contract, they should be. At one point they were optional because we hadn't computed them yet. Can you make a ticket for the work to fix the API contract and then make them non-optional in the UI code?

evansuslovich replied:
I'll write a ticket! sounds good!

pshapiro4broad commented on src/dataset-builder/DatasetBuilderUtils.ts:
```diff
@@ -114,54 +69,60 @@ export interface Cohort extends DatasetBuilderType {
+  participantCount?: number;
+}
+
+export type BuilderSettings = {
```
It looks like this type and `DatasetModel` aren't being used anywhere except in test code. Can you show me where this is used in the UI code? If they're not used, I think the fix here is to replace the test code usages with the API types.

evansuslovich replied:
I removed `DatasetModel`. The only reason I made `DatasetModel` was because `SnapshotBuilderDomainOption` removed `conceptCount` and `participantCount`.

rjohanek commented on src/dataset-builder/CohortEditor.ts:
```diff
@@ -54,23 +55,25 @@ const addCriteriaText = 'Add criteria';
+                  {
+                    // Create a "cohort" to get the count of participants for this criteria on its own.
+                    criteriaGroups: [{ criteria: [criteria], name: '', meetAll: true, mustMeet: true }],
+                    name: '',
```
I know this isn't your change, but I'm curious why is no name for the group or cohort saved?

evansuslovich replied:
Ya, this is a good question, let me look into this.

pshapiro4broad replied:
This API call is used to generate the `count` value for the criteria display. The `getCounts` API ignores names completely (it only looks at criteria) so they're not needed, and in this case there is no cohort name or cohort group name since we're making a fake one here just so we can get a rollup count separate from the overall cohort count.

rjohanek commented on src/dataset-builder/DatasetBuilderUtils.ts:
```diff
@@ -114,54 +69,60 @@ export interface Cohort extends DatasetBuilderType {
+export interface DomainOption extends SnapshotBuilderOption {
+  kind: 'domain';
+  root: SnapshotBuilderConcept;
+  conceptCount?: number;
```
what does concept count represent? The number of concepts or the number of participants

evansuslovich replied:
This may be wrong, but I believe `conceptCount` represents the amount of concepts that are in the DomainOption

pshapiro4broad replied:
I believe it's the number of concepts in the domain that have at least one entry in that domain's occurrence table (so there's at least one `person` that maps to that concept).

rjohanek commented:
looks good. I like the distinction between the UI and API types. I just have a few questions.

> ### Jira Ticket: https://broadworkbench.atlassian.net/browse/DC-1024
> ## Summary of changes:
> `DataRepo` adjustments:
> 
> * Removed `name` field from `Criteria`
> * renamed `getCounts` function to `getSnapshotBuilderCount`
> * Removed `id` from `SnapshotBuilderFeatureValueGroup`

Why remove id from this?

> * Renamed `AnyCriteria` to `AnySnapshotBuilderCriteria`
> * Created`SnapshotAccessRequestResponse` type
> * Removed `convertDatasetParticipantCountRequest` from `convertDatasetAccessRequest`
>   
>   * These conversion will be made before the request is passed in.
> 
> Types moved from `DatasetBuilderUtils` to`DataRepo` (and):
> 
> * `DatasetBuilderType`
> * `SnapshotBuilderParentConcept`
> * Renamed `GetConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> * Renamed `GetConceptHierarchyResponse `to `SnapshotBuilderGetConceptHierarchyResponse `
> * Renamed `SearchConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> * Renamed `DatasetParticipantCountResponse` to `SnapshotBuilderCountResponse`
> * Renamed `DatasetAccessRequestApi` to `DatasetAccessRequestApi`

So it seems like the convention for the naming of API types is to add SnapshotBuilder to the name. Why leave this one with API?

> * Renamed `DatasetRequestApi` to `SnapshotBuilderRequest`
> * Renamed `ValueSetApi` to `SnapshotBuilderFeatureValueGroup`

Is there a difference between a ValueSet and a Feature Value Group?

> * Renamed `CohortApi` to `SnapshotBuilderCohort`
> * Renamed `CriteriaGroupApi` to `SnapshotBuilderCriteriaGroup`
> * Renamed `ProgramDataListCriteriaApi`  to `SnapshotBuilderProgramDataListCriteria`
> * Renamed `ProgramDataRangeCriteriaApi` to `SnapshotBuilderProgramDataRangeCriteria`
> * Renamed `DomainCriteriaApi` to `SnapshotBuilderDomainCriteria`

> * Renamed `CriteriaApi` to `SnapshotBuilderCriteria`, removed `name` and `count`

Why remove count from the Criteria type? Don't we need this?

> Moved from `DataRepo` to `DatasetBuilderUtils`:
> 
> * `PrepackagedConceptSet`
> 
> Types created in `DatasetBuilderUtils`:
> 
> * `BuilderSettings`
> * `DatasetModel`
> * Renamed `DomainCriteria to `ProgramDomainCriteria`

Why add Program to this name? What's the reason for the distinction with the Program Data Criteria (list and range) and the Domain Criteria? Is this to bring more similarity across the three types?

pshapiro4broad commented on src/dataset-builder/DatasetBuilder.ts:
```diff
@@ -481,15 +483,20 @@ const RequestAccessModal = (props: RequestAccessModalProps) => {
-                },
-              });
+              .createSnapshotRequest(
+                convertDatasetAccessRequest({
```
This code is creating a object that's only used here only to "convert" it to another type. Why is the intermediate type needed? I would instead write a `create` function that creates the type directly, rather than converting.

evansuslovich commented:
> looks good. I like the distinction between the UI and API types. I just have a few questions.
> 
> > ### Jira Ticket: https://broadworkbench.atlassian.net/browse/DC-1024
> > ## Summary of changes:
> > `DataRepo` adjustments:
> > 
> > * Removed `name` field from `Criteria`
> > * renamed `getCounts` function to `getSnapshotBuilderCount`
> > * Removed `id` from `SnapshotBuilderFeatureValueGroup`
> 
> Why remove id from this?
> 
> > * Renamed `AnyCriteria` to `AnySnapshotBuilderCriteria`
> > * Created`SnapshotAccessRequestResponse` type
> > * Removed `convertDatasetParticipantCountRequest` from `convertDatasetAccessRequest`
> >   
> >   * These conversion will be made before the request is passed in.
> > 
> > Types moved from `DatasetBuilderUtils` to`DataRepo` (and):
> > 
> > * `DatasetBuilderType`
> > * `SnapshotBuilderParentConcept`
> > * Renamed `GetConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> > * Renamed `GetConceptHierarchyResponse `to `SnapshotBuilderGetConceptHierarchyResponse `
> > * Renamed `SearchConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> > * Renamed `DatasetParticipantCountResponse` to `SnapshotBuilderCountResponse`
> > * Renamed `DatasetAccessRequestApi` to `DatasetAccessRequestApi`
> 
> So it seems like the convention for the naming of API types is to add SnapshotBuilder to the name. Why leave this one with API?
> 
> > * Renamed `DatasetRequestApi` to `SnapshotBuilderRequest`
> > * Renamed `ValueSetApi` to `SnapshotBuilderFeatureValueGroup`
> 
> Is there a difference between a ValueSet and a Feature Value Group?
> 
> > * Renamed `CohortApi` to `SnapshotBuilderCohort`
> > * Renamed `CriteriaGroupApi` to `SnapshotBuilderCriteriaGroup`
> > * Renamed `ProgramDataListCriteriaApi`  to `SnapshotBuilderProgramDataListCriteria`
> > * Renamed `ProgramDataRangeCriteriaApi` to `SnapshotBuilderProgramDataRangeCriteria`
> > * Renamed `DomainCriteriaApi` to `SnapshotBuilderDomainCriteria`
> 
> > * Renamed `CriteriaApi` to `SnapshotBuilderCriteria`, removed `name` and `count`
> 
> Why remove count from the Criteria type? Don't we need this?
> 
> > Moved from `DataRepo` to `DatasetBuilderUtils`:
> > 
> > * `PrepackagedConceptSet`
> > 
> > Types created in `DatasetBuilderUtils`:
> > 
> > * `BuilderSettings`
> > * `DatasetModel`
> > * Renamed `DomainCriteria to `ProgramDomainCriteria`
> 
> Why add Program to this name? What's the reason for the distinction with the Program Data Criteria (list and range) and the Domain Criteria? Is this to bring more similarity across the three types?

Yes, the other types in `DatasetBuilderUtils` amongst criterias are `ProgramDataRangeCriteria` and `ProgramDataListCriteria` and now `ProgramDomainCriteria` instead of `DomainCriteria`. So basically, just add more similarity across the three types.

pshapiro4broad commented on src/dataset-builder/CohortEditor.ts:
```diff
@@ -54,23 +55,25 @@ const addCriteriaText = 'Add criteria';
-              ],
-            })
+            .getSnapshotBuilderCount(
+              convertDatasetParticipantCountRequest({
+                cohorts: [
+                  {
+                    // Create a "cohort" to get the count of participants for this criteria on its own.
+                    criteriaGroups: [{ criteria: [criteria], name: '', meetAll: true, mustMeet: true }],
+                    name: '',
+                  },
+                ],
+              })
```
Similar to `convertDatasetAccessRequest`, I think this is no longer a `convert` operation but a `create` one, e.g., as a function that accepts a `cohorts` argument directly
```suggestion
              createDatasetParticipantCountRequest(
                [
                  {
                    // Create a "cohort" to get the count of participants for this criteria on its own.
                    criteriaGroups: [{ criteria: [criteria], name: '', meetAll: true, mustMeet: true }],
                    name: '',
                  },
                ])
```

pshapiro4broad reviewed (commented):
Looks OK except for some issues with how `participantCount` and `conceptCount` are used

evansuslovich commented:
> > looks good. I like the distinction between the UI and API types. I just have a few questions.
> > > ### Jira Ticket: https://broadworkbench.atlassian.net/browse/DC-1024
> > > ## Summary of changes:
> > > `DataRepo` adjustments:
> > > 
> > > * Removed `name` field from `Criteria`
> > > * renamed `getCounts` function to `getSnapshotBuilderCount`
> > > * Removed `id` from `SnapshotBuilderFeatureValueGroup`
> > 
> > 
> > Why remove id from this?
I removed id because it's not required or used in `SnapshotBuilderFeatureValueGroup` in JDR 
<img width="525" alt="Screenshot 2024-05-01 at 3 40 37 PM" src="https://github.com/DataBiosphere/terra-ui/assets/63474660/4e9cc365-d273-4873-8f4b-c3e77ca3213b">



> > > * Renamed `AnyCriteria` to `AnySnapshotBuilderCriteria`
> > > * Created`SnapshotAccessRequestResponse` type
> > > * Removed `convertDatasetParticipantCountRequest` from `convertDatasetAccessRequest`
> > >   
> > >   * These conversion will be made before the request is passed in.
> > > 
> > > Types moved from `DatasetBuilderUtils` to`DataRepo` (and):
> > > 
> > > * `DatasetBuilderType`
> > > * `SnapshotBuilderParentConcept`
> > > * Renamed `GetConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> > > * Renamed `GetConceptHierarchyResponse `to `SnapshotBuilderGetConceptHierarchyResponse `
> > > * Renamed `SearchConceptsResponse` to `SnapshotBuilderGetConceptsResponse`
> > > * Renamed `DatasetParticipantCountResponse` to `SnapshotBuilderCountResponse`
> > > * Renamed `DatasetAccessRequestApi` to `DatasetAccessRequestApi`
> > 
> > 
> > So it seems like the convention for the naming of API types is to add SnapshotBuilder to the name. Why leave this one with API?
> > > * Renamed `DatasetRequestApi` to `SnapshotBuilderRequest`
> > > * Renamed `ValueSetApi` to `SnapshotBuilderFeatureValueGroup`
> > 
> > 
> > Is there a difference between a ValueSet and a Feature Value Group?
> > > * Renamed `CohortApi` to `SnapshotBuilderCohort`
> > > * Renamed `CriteriaGroupApi` to `SnapshotBuilderCriteriaGroup`
> > > * Renamed `ProgramDataListCriteriaApi`  to `SnapshotBuilderProgramDataListCriteria`
> > > * Renamed `ProgramDataRangeCriteriaApi` to `SnapshotBuilderProgramDataRangeCriteria`
> > > * Renamed `DomainCriteriaApi` to `SnapshotBuilderDomainCriteria`
> > 
> > 
> > > * Renamed `CriteriaApi` to `SnapshotBuilderCriteria`, removed `name` and `count`
> > 
> > 
> > Why remove count from the Criteria type? Don't we need this?
> > > Moved from `DataRepo` to `DatasetBuilderUtils`:
> > > 
> > > * `PrepackagedConceptSet`
> > > 
> > > Types created in `DatasetBuilderUtils`:
> > > 
> > > * `BuilderSettings`
> > > * `DatasetModel`
> > > * Renamed `DomainCriteria to `ProgramDomainCriteria`
> > 
> > 
> > Why add Program to this name? What's the reason for the distinction with the Program Data Criteria (list and range) and the Domain Criteria? Is this to bring more similarity across the three types?
> 
> Yes, the other types in `DatasetBuilderUtils`in the Criteria are `ProgramDataRangeCriteria` and `ProgramDataListCriteria` and now `ProgramDomainCriteria` instead of `DomainCriteria`. So basically, just add more similarity across the three types.

rjohanek reviewed (approved):
Thanks for answering my questions!

evansuslovich committed:
replace convertDatasetAccessRequest to createDatasetAccessRequest

evansuslovich committed:
replace convertDatasetParticipantCountRequest with createSnapshotBuilderCountRequest

evansuslovich committed:
Merge branch 'dev' into es/dc-1024-remove-name-field-from-criteria

evansuslovich committed:
removed DatasetModel from DatasetBuilderUtils

evansuslovich committed:
rename DatasetRequest to SnapshotBuilderRequest

evansuslovich committed:
make datasetConceptSets non-optional
